### PR TITLE
Sites of backend user should be reset when not valid

### PR DESCRIPTION
**Pull request by timoschmidt**
- Reinitialize Site in BackendModule when invalid site was retrieved from backend users uc data.
- Improved error message when no site is initialized.

Fixes: #142

### CODE REVIEW

File: Classes/Util.php [hunk 1]
```diff
@@ -716,5 +716,21 @@ public static function getAllowedPageTypes($pageId)
...
     }
+
+    /**
+     * Method to check if a page exists.
+     *
+     * @param integer $pid
+     *
+     * @return bool
+     */
+    public static function hasPage($pid) {
```
Comment (irnnr) [inline Classes/Util.php]: Can we call this pageExists()

Comment (irnnr) [inline Classes/Util.php]: also, please try to avoid abbreviations $pid -> $pageId

File: Resources/Private/Layouts/Default.html [hunk 2]
```diff
@@ -7,7 +7,19 @@
...
 				<div class="left">
-					<solr:backend.menu.SiteSelectorMenu name="site" />
+
+					<solr:backend.sites>
+						<f:if condition="{hasSites}">
+							<f:then>
+								<solr:backend.menu.SiteSelectorMenu name="site" />
+							</f:then>
+							<f:else>
+								No solr connections initialized!
```
Comment (irnnr) [inline Resources/Private/Layouts/Default.html]: When talking about the Solr server, please use uppercase S "Solr". When talking about the extension, it's EXT:solr with lowercase s.

File: Classes/Controller/AdministrationController.php [hunk 3]
```diff
@@ -105,7 +106,6 @@ public function processRequest(
         ResponseInterface $response
     ) {
         $this->moduleData = $this->moduleDataStorageService->loadModuleData();
-
```
Comment (irnnr) [inline Classes/Controller/AdministrationController.php]: I really like my whitespace :)

File: Classes/Controller/AdministrationController.php [hunk 4]
```diff
@@ -166,7 +174,11 @@ protected function invokeModuleController()
         $request->setControllerExtensionName(ucfirst($activeModuleDescription['extensionKey']));
         $request->setControllerName($activeModuleDescription['controller'] . 'Module');
         $request->setControllerActionName('index');
-        $request->setArgument('site', $this->site);
+
+        if( !is_null($this->site) ) {
```
Comment (irnnr) [inline Classes/Controller/AdministrationController.php]: Is this PSR-2? You can setup PhpStorm to check/correct/produce PSR-2 compliant code automatically.

File: Classes/Controller/AdministrationController.php [hunk 5]
```diff
@@ -265,5 +274,27 @@ protected function forwardToModule($module, $moduleAction = '')
 
         $this->forward('index');
     }
+
+    /**
+     * @param $site
+     */
+    protected function storeSiteAndResetCore($site)
```
Comment (irnnr) [inline Classes/Controller/AdministrationController.php]: would prefer to call this setSite...

File: Classes/Site.php [hunk 6]
```diff
@@ -162,6 +170,19 @@ public static function getAvailableSites()
     }
 
     /**
+     * Returns the first available Site.
+     *
+     * @param boolean $stopOnInvalidSite
```
Comment (irnnr) [inline Classes/Site.php]: What's the intention of this new parameter? I don't see it used....

File: Classes/ViewHelpers/Backend/SitesViewHelper.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
+<?php
+namespace ApacheSolrForTypo3\Solr\ViewHelpers\Backend;
+
+/***************************************************************
+ *  Copyright notice
+ *
+ *  (c) 2013-2015 Ingo Renner <ingo@typo3.org>
```
Comment (irnnr) [inline Classes/ViewHelpers/Backend/SitesViewHelper.php]: Feel free to use your name

File: Resources/Private/Layouts/Default.html (hunk 2 above)
Comment (irnnr) [inline Resources/Private/Layouts/Default.html]: In general I think I have to check the empty state again - what is shown the user hits the module the first time. Would be nice to offer a little message and a button to initialize the sites.

File: Resources/Private/Layouts/Default.html [hunk 8]
```diff
@@ -23,12 +35,32 @@
...
+								<p>Please check the following:</p>
+								<ul>
+									<li>Your page has a domain record.</li>
+									<li>Your site has a page, marked as root page.</li>
+									<li>Your page has included the solr typoscript configuration.</li>
+									<li>You've initialized the solr connections in the backend with the menu item "Initialize solr connections"</li>
+									<li>Afterwards please check if the reports module is showing any errors.</li>
+								</ul>
+
+								<p>When all of these steps are done check this module again.</p>
```
Comment (irnnr) [inline Resources/Private/Layouts/Default.html]: :+1:

File: Classes/Util.php (hunk 1 above)
Comment (timoschmidt) [inline Classes/Util.php]: Changed both

File: Resources/Private/Layouts/Default.html (hunk 2 above)
Comment (timoschmidt) [inline Resources/Private/Layouts/Default.html]: Also changed

File: Classes/Controller/AdministrationController.php (hunk 3 above)
Comment (timoschmidt) [inline Classes/Controller/AdministrationController.php]: Also changed

File: Classes/Controller/AdministrationController.php (hunk 5 above)
Comment (timoschmidt) [inline Classes/Controller/AdministrationController.php]: Also changed

File: Classes/Site.php (hunk 6 above)
Comment (timoschmidt) [inline Classes/Site.php]: The intension is if the Exception from the constructor is bubbled up or not. Normally i would say it is good the the exception get's thrown, but when getting the collection in a loop it breaks the look, which is not allways wanted. Sometimes you just want to skip this item, which is the default now.

File: Classes/ViewHelpers/Backend/SitesViewHelper.php (hunk 7 above)
Comment (timoschmidt) [inline Classes/ViewHelpers/Backend/SitesViewHelper.php]: Done

**Pull request merged**